### PR TITLE
GA's main skeleton

### PR DESCRIPTION
**Pull request by galyfray**
## What has been done
 - Implementation of the `ConstraintValidator` Interface
   - Added all the constraint I thought of.
 - Implementation of the Individual Abstract Class
   - Added the `ECVRPSolution` class with the following method implemented : 
     - get_instance
     - get_points
     - get_roads
     - get_fitness
     - is_valid
     - \_\_copy__
 - Added the `ECVRPInstance` class to holds most of the information
 -  Created tests for the various class and methods added

## What has not been done
 - The mutation and crossover methods were not done
 - The GA class itself
 - The system in charge of reading the benchmarks
   - should be part of the server
 The main reasons behind the unimplemented functionality are : 
 - I wanted to keep this PR in a manageable size.
 - I wanted to end this PR before the next sprint cycle. 

## Coverage
|Name  |  Stmts |  Miss | Cover  |
|-|-|-|-|
|||||
|TOTAL |    256  |    0 |  100%|

With 100% coverage and no red flag showed by the linters, I think this PR is up to the standard.

### CODE REVIEW

**Comment by galyfray** [comment]
the `Python application / build` failed because he couldn't update PIP ...

**Review by Nexowo — DISMISSED** [review]

File: src/cvrp/constraints_validators.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+from .individual import ConstraintValidator
+from .ecvrp import ECVRPSolution
+
+
+# pylint: disable=too-few-public-methods
+class BatteryTWValidator(ConstraintValidator[ECVRPSolution]):
+    """ Class in charge of checking the battery capacity is never exeded
```
[the remark this answers is not in the document]
Comment (galyfray) [reply src/cvrp/constraints_validators.py]: done

File: src/cvrp/ecvrp.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    def __init__(
+            self,
+            validators: list[ConstraintValidator],
+            solution: list[int],
+            instance: "ECVRPInstance"
+            ) -> None:
+        super().__init__(validators)
+        self._solution = solution
+        self.__instance = instance
+        self._roads = None
```
[the remark this answers is not in the document]
Comment (galyfray) [reply src/cvrp/ecvrp.py:54]: done

File: src/cvrp/ecvrp.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        self._fitness = None
```
[the remark this answers is not in the document]
Comment (galyfray) [reply src/cvrp/ecvrp.py:55]: done

**Review by m-aspro — CHANGES_REQUESTED** [review]
Good job! 
Just a little point to modify.

File: src/cvrp/constraints_validators.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+
+from .individual import ConstraintValidator
+from .ecvrp import ECVRPSolution
+
+
+# pylint: disable=too-few-public-methods
+class BatteryTWValidator(ConstraintValidator[ECVRPSolution]):
+    """
+    Class in charge of checking if the battery capacity is never exeded
```
Comment (Axelvel) [inline src/cvrp/constraints_validators.py:36]: Typo in the docstring. 
```suggestion
    Class in charge of checking if the battery capacity is never exceded
```

**Comment by galyfray** [comment]
> Good job! Just a little point to modify.

Thanks. But, you did not mention what should be changed in your review.

File: src/cvrp/constraints_validators.py (hunk 4 above)
Comment (m-aspro) [inline src/cvrp/constraints_validators.py:36]: Could you specify about TW ? 
I mean by adding a comment which explain that is also a time window validator.
Thanks!

**Comment by m-aspro** [comment]
I added a comment by quoting the line, I think it has been skipped or cancel by error. 
I'll do it again right away, sorry for that! :)

File: src/cvrp/constraints_validators.py (hunk 4 above)
Comment (m-aspro) [reply src/cvrp/constraints_validators.py]: Oh you added it just after we discuss, so it's okay for me, I see no problem

**Review by m-aspro — APPROVED** [review]

**Review by KwassiSenam — APPROVED** [review]

**Pull request merged**